parser/parser.coffee

```diff
@@ -0,0 +1,118 @@
+scanner = require '../scanner/scanner'
+CustomError = require '../error/custom_error'
+
+Program = require '../entities/01_program'
+Block = require '../entities/02_block'
+ForStatement = require '../entities/03_for_statement'
+WhileStatement = require '../entities/03_while_statement'
+
+StartTokens = require './start_tokens'
+
+tokens = []
+
+module.exports = (scannerOutput) ->
+  tokens = scannerOutput
+  program = parseProgram()
+  match 'EOF'
+
+parseProgram = ->
+  new Program parseBlock
+
+parseBlock = ->
+  statements = []
+  loop
+    statements.push parseStatement()
+    match 'newline'
+    break unless at StartTokens.statement
+  new Block statements
+
+parseStatement = ->
+  if at 'for'
+    parseForLoop()
+  else if at 'while'
+    parseWhileLoop()
+  else
+    parseExpression()
+
+parseForLoop = ->
+  match 'for'
+  id = parseId()
```

Comment:
Good catch, Rachel. (1) Code shouldn't be pushed if it calls a function that is not defined, and (2) ID is a token, not a syntactic category, so I see no reason that a parsing function should be written for it. What is the intent here?